File: apps/web/src/server/auth/password.ts
import { randomBytes, scrypt, timingSafeEqual } from "node:crypto";
import { promisify } from "node:util";

const scryptAsync = promisify(scrypt);

const DEFAULT_PARAMS = {
  N: 16384,
  r: 8,
  p: 1,
  keyLength: 64,
  saltLength: 16
};

const parseParams = (segment: string) => {
  const parsed: Record<string, number> = {};
  for (const part of segment.split(",")) {
    const [key, value] = part.split("=");
    const numeric = Number(value);
    if (key && Number.isFinite(numeric)) parsed[key.trim()] = numeric;
  }
  return parsed;
};

export const hashPassword = async (password: string) => {
  const salt = randomBytes(DEFAULT_PARAMS.saltLength);
  const derivedKey = (await scryptAsync(password, salt, DEFAULT_PARAMS.keyLength, {
    cost: DEFAULT_PARAMS.N,
    blockSize: DEFAULT_PARAMS.r,
    parallelization: DEFAULT_PARAMS.p
  })) as Buffer;

  const params = `N=${DEFAULT_PARAMS.N},r=${DEFAULT_PARAMS.r},p=${DEFAULT_PARAMS.p}`;
  const saltEncoded = salt.toString("base64");
  const hashEncoded = derivedKey.toString("base64");
  return `$scrypt$${params}$${saltEncoded}$${hashEncoded}`;
};

export const verifyPassword = async (password: string, storedHash: string) => {
  if (!storedHash.startsWith("$scrypt$")) return false;
  const parts = storedHash.split("$");
  if (parts.length < 5) return false;

  const params = parseParams(parts[2] ?? "");
  const saltEncoded = parts[3];
  const hashEncoded = parts[4];
  if (!saltEncoded || !hashEncoded) return false;

  const salt = Buffer.from(saltEncoded, "base64");
  const expected = Buffer.from(hashEncoded, "base64");

  const derivedKey = (await scryptAsync(password, salt, expected.length, {
    cost: params.N ?? DEFAULT_PARAMS.N,
    blockSize: params.r ?? DEFAULT_PARAMS.r,
    parallelization: params.p ?? DEFAULT_PARAMS.p
  })) as Buffer;

  if (derivedKey.length !== expected.length) return false;
  return timingSafeEqual(derivedKey, expected);
};

File: apps/server/internal/modules/auth/password.go
package auth

import (
	"crypto/subtle"
	"encoding/base64"
	"strconv"
	"strings"

	"golang.org/x/crypto/scrypt"
)

const (
	defaultScryptN = 16384
	defaultScryptR = 8
	defaultScryptP = 1
)

func VerifyPassword(password string, storedHash string) bool {
	if !strings.HasPrefix(storedHash, "$scrypt$") {
		return false
	}

	parts := strings.Split(storedHash, "$")
	if len(parts) < 5 {
		return false
	}

	params := parseParams(parts[2])
	saltEncoded := parts[3]
	hashEncoded := parts[4]
	if saltEncoded == "" || hashEncoded == "" {
		return false
	}

	salt, err := base64.StdEncoding.DecodeString(saltEncoded)
	if err != nil {
		return false
	}
	expected, err := base64.StdEncoding.DecodeString(hashEncoded)
	if err != nil {
		return false
	}

	derived, err := scrypt.Key([]byte(password), salt, params.N, params.R, params.P, len(expected))
	if err != nil {
		return false
	}

	return subtle.ConstantTimeCompare(derived, expected) == 1
}

type scryptParams struct {
	N int
	R int
	P int
}

func parseParams(segment string) scryptParams {
	params := scryptParams{N: defaultScryptN, R: defaultScryptR, P: defaultScryptP}
	for _, part := range strings.Split(segment, ",") {
		pieces := strings.SplitN(strings.TrimSpace(part), "=", 2)
		if len(pieces) != 2 {
			continue
		}
		value, err := strconv.Atoi(strings.TrimSpace(pieces[1]))
		if err != nil {
			continue
		}
		switch pieces[0] {
		case "N":
			params.N = value
		case "r":
			params.R = value
		case "p":
			params.P = value
		}
	}
	return params
}
